File: forms/workplace-safety-assessment/back-end-with-loco/src/engine/safety_grader.rs
//! Safety grader module.

// Workplace Safety Assessment grader. Pure functions: take an
// `AssessmentData` object, return the overall outcome, the per-category
// findings tally, the list of rules that fired, and the additional flags.
//
// Outcome rules:
//   any rule fired at grade 4         -> 'critical'
//   any rule fired at grade 3 (no 4s) -> 'major'
//   any rule fired at grade 2 (no 3-4) -> 'minor'
//   only grade-1 rules                -> 'compliant'
//
// Rules that score 0 are unanswered and excluded entirely from the grading
// totals (so a partially-completed audit doesn't auto-fail).

use std::collections::BTreeMap;

use super::flagged_issues::detect_additional_flags;
use super::rules::all_rules;
use super::types::{
    AssessmentData, CategoryFindings, FiredRule, GradingResult, Outcome, SeverityGrade,
};
use super::utils::grade_to_finding_level;

/// Determine the worst outcome from a set of fired rules.
pub fn highest_outcome(fired_rules: &[FiredRule]) -> Outcome {
    let mut worst: SeverityGrade = 1;
    for r in fired_rules {
        if r.grade > worst {
            worst = r.grade;
        }
    }
    grade_to_finding_level(worst)
}

/// Pure function: grades a Workplace Safety Assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut findings_by_category: BTreeMap<String, CategoryFindings> = BTreeMap::new();
    let mut answered_count: u32 = 0;

    for rule in all_rules() {
        let grade_val = (rule.evaluate)(data);
        if grade_val == 0 {
            continue; // unanswered
        }
        answered_count += 1;

        fired_rules.push(FiredRule {
            id: rule.id.to_string(),
            category: rule.category.to_string(),
            description: rule.description.to_string(),
            grade: grade_val,
        });

        let bucket = findings_by_category
            .entry(rule.category.to_string())
            .or_insert_with(|| CategoryFindings {
                category: rule.category.to_string(),
                ..CategoryFindings::default()
            });
        bucket.total += 1;
        match grade_val {
            4 => bucket.critical += 1,
            3 => bucket.major += 1,
            2 => bucket.minor += 1,
            _ => bucket.compliant += 1,
        }
    }

    let outcome = if answered_count == 0 {
        "compliant".to_string()
    } else {
        highest_outcome(&fired_rules)
    };

    let additional_flags = detect_additional_flags(data);

    GradingResult {
        outcome,
        findings_by_category,
        fired_rules,
        additional_flags,
        answered_count,
        timestamp,
    }
}

File: forms/workplace-safety-assessment/front-end-with-svelte/src/lib/engine/safety-grader.ts
import type {
	AssessmentData,
	Outcome,
	FiredRule,
	CategoryFindings,
	SeverityGrade,
	GradingResult
} from './types';
import { safetyRules } from './safety-rules';
import { gradeToFindingLevel } from './utils';
import { detectAdditionalFlags } from './flagged-issues';

// Workplace Safety Assessment grader. Pure functions: take an `AssessmentData`
// object, return the overall outcome, the per-category findings tally, and the
// list of rules that fired (compliant + non-compliant).
//
// Outcome rules:
//   any rule fired at grade 4         -> 'critical'
//   any rule fired at grade 3 (no 4s) -> 'major'
//   any rule fired at grade 2 (no 3-4) -> 'minor'
//   only grade-1 rules                 -> 'compliant'
//
// Rules that score 0 are unanswered and excluded entirely from the totals (so a
// partially-completed audit does not auto-fail).

/** Determine the worst outcome from a set of fired rules. */
export function highestOutcome(firedRules: FiredRule[]): Outcome {
	let worst = 1;
	for (const r of firedRules) {
		if (r.grade > worst) worst = r.grade;
	}
	return gradeToFindingLevel(worst);
}

/** Evaluate the workplace safety audit checklist against the supplied data. */
export function gradeSafety(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	const findingsByCategory: Record<string, CategoryFindings> = {};
	let answeredCount = 0;

	for (const rule of safetyRules) {
		let grade: number;
		try {
			grade = rule.evaluate(data);
		} catch (e) {
			console.warn(`Safety rule ${rule.id} evaluation failed:`, e);
			continue;
		}

		if (grade === 0) continue; // unanswered
		answeredCount++;

		const g = grade as SeverityGrade;
		firedRules.push({
			id: rule.id,
			category: rule.category,
			description: rule.description,
			grade: g
		});

		if (!findingsByCategory[rule.category]) {
			findingsByCategory[rule.category] = {
				category: rule.category,
				compliant: 0,
				minor: 0,
				major: 0,
				critical: 0,
				total: 0
			};
		}
		const bucket = findingsByCategory[rule.category];
		bucket.total++;
		if (g === 4) bucket.critical++;
		else if (g === 3) bucket.major++;
		else if (g === 2) bucket.minor++;
		else bucket.compliant++;
	}

	const outcome: Outcome = answeredCount === 0 ? 'compliant' : highestOutcome(firedRules);

	return {
		outcome,
		findingsByCategory,
		firedRules,
		additionalFlags: detectAdditionalFlags(data),
		answeredCount,
		timestamp: new Date().toISOString()
	};
}
